lib/uploader.ts

```diff
@@ -234,40 +234,88 @@ export class Uploader {
 	batchUpload(
 		destination: string,
 		files: (File|FileSystemEntry)[],
-		callback?: (nodes: Array<File|IDirectory>,
-		currentPath: string) => Promise<Array<File|IDirectory>|false>,
+		callback?: (nodes: Array<File|IDirectory>, currentPath: string) => Promise<Array<File|IDirectory>|false>,
 	): PCancelable<Upload[]> {
 		const rootFolder = new Directory('', files)
 		if (!callback) {
 			callback = async (files: Array<File|Directory>) => files
 		}
 
-		try {
-			// Increase concurrency to 4 to keep 3 parallel uploads as one if blocked by the directory meta-upload
-			this._jobQueue.concurrency += 1
+		return new PCancelable(async (resolve, reject, onCancel) => {
+			// create a meta upload to ensure all ongoing child requests are listed
+			const upload = new Upload(`${this.root.replace(/\/$/, '')}/${destination.replace(/^\//, '')}`, false, 0, rootFolder)
+			upload.status = UploadStatus.UPLOADING
+			this._uploadQueue.push(upload)
+			try {
+				// Create the promise for the virtual root directory
+				const promise = this.uploadDirectory(destination, rootFolder, callback, davGetClient(this.root))
+				// Make sure to cancel it when requested
+				onCancel(() => promise.cancel())
+				// await the uploads and resolve with "finished" status
+				const uploads = await promise
+				upload.status = UploadStatus.FINISHED
+				resolve(uploads)
+			} catch (error) {
+				logger.error('Error in batch upload', { error })
+				upload.status = UploadStatus.FAILED
```

Comment:
This broke the cancellation detection.
https://github.com/nextcloud/server/blob/b06f5ba4c47450f355a8903c1a93ac68e8c6cfc2/apps/files/src/views/FilesList.vue#L658-L661

Reply:
https://github.com/nextcloud-libraries/nextcloud-upload/pull/1614